<a href="https://colab.research.google.com/github/lnwprameza007/bsc_dpdm24/blob/main/Frequent_Patterns_(Association_Rules)_(dpdm24).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [54]:
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns

In [55]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [56]:
path = '/content/drive/MyDrive/bsc_dpdm24/data'

In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
# prompt: create a list of CustomerID and count and sum in count row
# Group by CustomerID and count the occurrences of each CustomerID
customer_counts = data.groupby('CustomerID')['CustomerID'].count()

# Create a DataFrame from the customer counts
customer_counts_df = pd.DataFrame({'CustomerID': customer_counts.index, 'Count': customer_counts.values})


# Calculate the sum of the 'Count' column
total_count = customer_counts_df['Count'].sum()

# Add the total count as a new row to the DataFrame
new_row = pd.DataFrame({'CustomerID': ['Total'], 'Count': [total_count]})
customer_counts_df = pd.concat([customer_counts_df, new_row], ignore_index=True)


customer_counts_df

CustomerID   Count
0       12346.0       2
1       12347.0     182
2       12348.0      31
3       12349.0      73
4       12350.0      17
...         ...     ...
4368    18281.0       7
4369    18282.0      13
4370    18283.0     756
4371    18287.0      70
4372      Total  406829

[4373 rows x 2 columns]

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [ ]:
!ls

drive  sample_data


In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=611df96c3a4e0b20a89090a3586d48d0e3740b0dbb2135888cdedcd39e4153b2
  Stored in directory: /root/.cache/pip/wheels/77/3d/a6/d317a6fb32be58a602b1e8c6b5d6f31f79322da554cad2a5ea
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
list(apriori(transactions,min_support = 0.5))

[]

In [ ]:
list(apriori(transactions,min_support = 0.1))

[]

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.03))
frequentItemsets

[RelationRecord(items=frozenset({'20712'}), support=0.033667953667953665, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.033667953667953665, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.03254826254826255, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.03254826254826255, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04054054054054054, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04054054054054054, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.062084942084942085, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.062084942084942085, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.04003861003861004, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozense

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
# มีทั้ง 4 กฎ

In [ ]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [ ]:
dir(frequentItemsets[0][0]) # ศึกษาโครงสร้างก่อนที่จะแปลงเป็น list ## dir ใช้ดู properties ของ Itemset

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
len(frequentItemsets[-1][0])

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216,
ordered_statistics=
[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}),items_add=frozenset({'85099B'}),
confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}),
confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
data[data['StockCode']=='85099B']

InvoiceNo StockCode              Description  Quantity  \
177       536386    85099B  JUMBO BAG RED RETROSPOT       100   
234       536390    85099B  JUMBO BAG RED RETROSPOT       100   
533       536409    85099B  JUMBO BAG RED RETROSPOT         2   
790       536464    85099B  JUMBO BAG RED RETROSPOT         1   
1069      536522    85099B  JUMBO BAG RED RETROSPOT         1   
...          ...       ...                      ...       ...   
538603    581412    85099B  JUMBO BAG RED RETROSPOT         3   
538913    581421    85099B  JUMBO BAG RED RETROSPOT         9   
539338    581439    85099B  JUMBO BAG RED RETROSPOT         1   
540434    581486    85099B  JUMBO BAG RED RETROSPOT        10   
541536    581498    85099B  JUMBO BAG RED RETROSPOT         5   

             InvoiceDate  UnitPrice  CustomerID         Country  
177     12/01/2010 09:57       1.65     16029.0  United Kingdom  
234     12/01/2010 10:19       1.65     17511.0  United Kingdom  
533     12/01/2010 11:45       1.95     17908.0  United Kingdom  
790     12/01/2010 12:23       1.95     17968.0  United Kingdom  
1069    12/01/2010 12:49       1.95     15012.0  United Kingdom  
...                  ...        ...         ...             ...  
538603  12/08/2011 14:38       2.08     14415.0  United Kingdom  
538913  12/08/2011 15:23       2.08     15898.0  United Kingdom  
539338  12/08/2011 16:30       4.13         NaN  United Kingdom  
540434  12/09/2011 09:38       2.08     17001.0  United Kingdom  
541536  12/09/2011 10:26       4.13         NaN  United Kingdom  

[2159 rows x 8 columns]

In [ ]:
data[data['StockCode']=='22386']

InvoiceNo StockCode              Description  Quantity  \
86        536378     22386  JUMBO BAG PINK POLKADOT        10   
789       536464     22386  JUMBO BAG PINK POLKADOT         1   
1428      536542     22386  JUMBO BAG PINK POLKADOT        30   
2427      536591     22386  JUMBO BAG PINK POLKADOT         1   
2712      536592     22386  JUMBO BAG PINK POLKADOT         3   
...          ...       ...                      ...       ...   
534742    581167     22386  JUMBO BAG PINK POLKADOT        20   
537138    581238     22386  JUMBO BAG PINK POLKADOT         5   
539585    581439     22386  JUMBO BAG PINK POLKADOT         2   
541421    581498     22386  JUMBO BAG PINK POLKADOT         1   
541806    581579     22386  JUMBO BAG PINK POLKADOT        10   

             InvoiceDate  UnitPrice  CustomerID         Country  
86      12/01/2010 09:37       1.95     14688.0  United Kingdom  
789     12/01/2010 12:23       1.95     17968.0  United Kingdom  
1428    12/01/2010 14:11       1.95     16456.0  United Kingdom  
2427    12/01/2010 16:57       1.95     14606.0  United Kingdom  
2712    12/01/2010 17:06       4.21         NaN  United Kingdom  
...                  ...        ...         ...             ...  
534742  12/07/2011 14:52       2.08     13534.0  United Kingdom  
537138  12/08/2011 10:53       4.13         NaN  United Kingdom  
539585  12/08/2011 16:30       4.13         NaN  United Kingdom  
541421  12/09/2011 10:26       4.13         NaN  United Kingdom  
541806  12/09/2011 12:19       1.79     17581.0  United Kingdom  

[1251 rows x 8 columns]

# HW 3 จาก Data Reduced_Marketbasket (Group of 3)
1. หาพฤติกรรมผู้บริโภค (groupby customer id) แต่ละทวีปว่าแตกต่างกันยังไง
2. การวิเคราะห์ความสัมพันธ์ของสินค้า (Market Basket Analysis) โดยใช้ Association Rules อย่าง Apriori Algorithm เพื่อดูสินค้าที่ลูกค้ามักซื้อร่วมกัน จะได้ออกโปรโมชั่นสำหรับสินค้าคู่ หรือเปนตัวช่วยในการจัดวางสินค้าให้ใกล้กันเพื่อเพิ่มยอดขายในสินค้าตัวอื่นๆ

In [57]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [58]:
# check NaN
data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [59]:
countries = data['Country'].unique()
countries

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [60]:
country_to_continent = {
    'United Kingdom': 'EU',
    'France': 'EU',
    'Australia': 'OC',
    'Netherlands': 'EU',
    'Germany': 'EU',
    'Norway': 'EU',
    'EIRE': 'EU',
    'Switzerland': 'EU',
    'Spain': 'EU',
    'Poland': 'EU',
    'Portugal': 'EU',
    'Italy': 'EU',
    'Belgium': 'EU',
    'Lithuania': 'EU',
    'Japan': 'AS',
    'Iceland': 'EU',
    'Channel Islands': 'EU',
    'Denmark': 'EU',
    'Cyprus': 'EU',
    'Sweden': 'EU',
    'Finland': 'EU',
    'Austria': 'EU',
    'Greece': 'EU',
    'Hong Kong': 'AS',
    'Singapore': 'AS',
    'Lebanon': 'AS',
    'United Arab Emirates': 'AS',
    'Israel': 'AS',
    'Saudi Arabia': 'AS',
    'Czech Republic': 'EU',
    'Canada': 'NA',
    'Brazil': 'SA',
    'USA': 'NA',
    'European Community': 'EU',
    'Bahrain': 'AS',
    'Malta': 'EU',
    'RSA': 'AF',
    'Unspecified': 'Unknown'  # Treat 'Unspecified' as 'Unknown'
}

In [61]:
data['Legion'] = data['Country'].map(country_to_continent)
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Legion  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom     EU  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom     EU  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
...                  ...        ...         ...             ...    ...  
541904  12/09/2011 12:50       0.85     12680.0          France     EU  
541905  12/09/2011 12:50       2.10     12680.0          France     EU  
541906  12/09/2011 12:50       4.15     12680.0          France     EU  
541907  12/09/2011 12:50       4.15     12680.0          France     EU  
541908  12/09/2011 12:50       4.95     12680.0          France     EU  

[541909 rows x 9 columns]

In [62]:
data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
Legion              0
dtype: int64

In [63]:
data[data['Country'] == 'Unspecified']

InvoiceNo StockCode                          Description  Quantity  \
152712    549687     20685                DOORMAT RED RETROSPOT         2   
152713    549687     22691              DOORMAT WELCOME SUNRISE         2   
152714    549687     48116           DOORMAT MULTICOLOUR STRIPE         2   
152715    549687     21213          PACK OF 72 SKULL CAKE CASES        24   
152716    549687     21977   PACK OF 60 PINK PAISLEY CAKE CASES        24   
...          ...       ...                                  ...       ...   
498778    578539     22560           TRADITIONAL MODELLING CLAY        24   
498779    578539     23570     TRADITIONAL PICK UP STICKS GAME         12   
498780    578539     23571        TRADITIONAL NAUGHTS & CROSSES        12   
498781    578539     84992       72 SWEETHEART FAIRY CAKE CASES        24   
498782    578539     23309  SET OF 60 I LOVE LONDON CAKE CASES         24   

             InvoiceDate  UnitPrice  CustomerID      Country   Legion  
152712  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152713  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152714  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152715  04/11/2011 13:29       0.55     12363.0  Unspecified  Unknown  
152716  04/11/2011 13:29       0.55     12363.0  Unspecified  Unknown  
...                  ...        ...         ...          ...      ...  
498778  11/24/2011 14:55       1.25         NaN  Unspecified  Unknown  
498779  11/24/2011 14:55       1.25         NaN  Unspecified  Unknown  
498780  11/24/2011 14:55       1.65         NaN  Unspecified  Unknown  
498781  11/24/2011 14:55       0.55         NaN  Unspecified  Unknown  
498782  11/24/2011 14:55       0.55         NaN  Unspecified  Unknown  

[446 rows x 9 columns]

In [64]:
data[(data['Legion'] == 'Unknown')]

InvoiceNo StockCode                          Description  Quantity  \
152712    549687     20685                DOORMAT RED RETROSPOT         2   
152713    549687     22691              DOORMAT WELCOME SUNRISE         2   
152714    549687     48116           DOORMAT MULTICOLOUR STRIPE         2   
152715    549687     21213          PACK OF 72 SKULL CAKE CASES        24   
152716    549687     21977   PACK OF 60 PINK PAISLEY CAKE CASES        24   
...          ...       ...                                  ...       ...   
498778    578539     22560           TRADITIONAL MODELLING CLAY        24   
498779    578539     23570     TRADITIONAL PICK UP STICKS GAME         12   
498780    578539     23571        TRADITIONAL NAUGHTS & CROSSES        12   
498781    578539     84992       72 SWEETHEART FAIRY CAKE CASES        24   
498782    578539     23309  SET OF 60 I LOVE LONDON CAKE CASES         24   

             InvoiceDate  UnitPrice  CustomerID      Country   Legion  
152712  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152713  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152714  04/11/2011 13:29       7.95     12363.0  Unspecified  Unknown  
152715  04/11/2011 13:29       0.55     12363.0  Unspecified  Unknown  
152716  04/11/2011 13:29       0.55     12363.0  Unspecified  Unknown  
...                  ...        ...         ...          ...      ...  
498778  11/24/2011 14:55       1.25         NaN  Unspecified  Unknown  
498779  11/24/2011 14:55       1.25         NaN  Unspecified  Unknown  
498780  11/24/2011 14:55       1.65         NaN  Unspecified  Unknown  
498781  11/24/2011 14:55       0.55         NaN  Unspecified  Unknown  
498782  11/24/2011 14:55       0.55         NaN  Unspecified  Unknown  

[446 rows x 9 columns]

In [65]:
data['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [66]:
data = data[data['Legion'] != 'Unknown']
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Legion  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom     EU  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom     EU  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
...                  ...        ...         ...             ...    ...  
541904  12/09/2011 12:50       0.85     12680.0          France     EU  
541905  12/09/2011 12:50       2.10     12680.0          France     EU  
541906  12/09/2011 12:50       4.15     12680.0          France     EU  
541907  12/09/2011 12:50       4.15     12680.0          France     EU  
541908  12/09/2011 12:50       4.95     12680.0          France     EU  

[541463 rows x 9 columns]

In [67]:
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     134878
Country             0
Legion              0
dtype: int64

In [68]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Legion  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom     EU  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom     EU  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
...                  ...        ...         ...             ...    ...  
541904  12/09/2011 12:50       0.85     12680.0          France     EU  
541905  12/09/2011 12:50       2.10     12680.0          France     EU  
541906  12/09/2011 12:50       4.15     12680.0          France     EU  
541907  12/09/2011 12:50       4.15     12680.0          France     EU  
541908  12/09/2011 12:50       4.95     12680.0          France     EU  

[541463 rows x 9 columns]

In [ ]:
from apyori import apriori

In [ ]:
transactions = data.groupby(['CustomerID','Legion'])['StockCode'].apply(list)

transactions

CustomerID  Legion
12346.0     EU                                           [23166, 23166]
12347.0     EU        [85116, 22375, 71477, 22492, 22771, 22772, 227...
12348.0     EU        [84992, 22951, 84991, 84991, 21213, 21213, 226...
12349.0     EU        [23112, 23460, 21564, 21411, 21563, 22131, 221...
12350.0     EU        [21908, 22412, 79066K, 79191C, 22348, 84086C, ...
                                            ...                        
18280.0     EU        [82484, 22180, 22467, 22725, 22727, 22495, 223...
18281.0     EU        [22037, 22716, 22028, 23007, 23008, 23209, 22467]
18282.0     EU        [21270, 23187, 23295, 22089, 21108, 21109, 224...
18283.0     EU        [22356, 20726, 22384, 22386, 20717, 20718, 850...
18287.0     EU        [22755, 22754, 22753, 22756, 22758, 22757, 227...
Name: StockCode, Length: 4370, dtype: object

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.03))
frequentItemsets

[RelationRecord(items=frozenset({'15036'}), support=0.044393592677345536, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15036'}), confidence=0.044393592677345536, lift=1.0)]),
 RelationRecord(items=frozenset({'15056BL'}), support=0.032494279176201374, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15056BL'}), confidence=0.032494279176201374, lift=1.0)]),
 RelationRecord(items=frozenset({'15056N'}), support=0.042105263157894736, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15056N'}), confidence=0.042105263157894736, lift=1.0)]),
 RelationRecord(items=frozenset({'16156S'}), support=0.030892448512585814, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'16156S'}), confidence=0.030892448512585814, lift=1.0)]),
 RelationRecord(items=frozenset({'16161P'}), support=0.043020594965675056, ordered_statistics=[OrderedStatistic(items_base=frozenset(), item

In [ ]:
min_sup = 0.07

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'20725', '20728'}), support=0.07048054919908467, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '20728'}), confidence=0.07048054919908467, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'20728'}), confidence=0.5789473684210527, lift=5.25987525987526), OrderedStatistic(items_base=frozenset({'20728'}), items_add=frozenset({'20725'}), confidence=0.6403326403326404, lift=5.25987525987526)])
yes
RelationRecord(items=frozenset({'21733', '85123A'}), support=0.07139588100686499, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21733', '85123A'}), confidence=0.07139588100686499, lift=1.0), OrderedStatistic(items_base=frozenset({'21733'}), items_add=frozenset({'85123A'}), confidence=0.8125, lift=4.138257575757575), OrderedStatistic(items_base=frozenset({'85123A'}), items_add=frozenset({'21733'}), confidence=0.3636363636363636, lift=4.138257575757575

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Legion  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom     EU  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom     EU  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom     EU  
...                  ...        ...         ...             ...    ...  
541904  12/09/2011 12:50       0.85     12680.0          France     EU  
541905  12/09/2011 12:50       2.10     12680.0          France     EU  
541906  12/09/2011 12:50       4.15     12680.0          France     EU  
541907  12/09/2011 12:50       4.15     12680.0          France     EU  
541908  12/09/2011 12:50       4.95     12680.0          France     EU  

[541463 rows x 9 columns]

In [69]:
# prompt: find most invoice by country

import pandas as pd

# Assuming 'data' DataFrame is already loaded as shown in your provided code

invoice_counts_by_country = data.groupby('Country')['InvoiceNo'].count()
most_invoices_country = invoice_counts_by_country.idxmax()
print(f"The country with the most invoices is: {most_invoices_country}")

The country with the most invoices is: United Kingdom


In [72]:
# prompt: แบ่งเวลาเช้า กลางวันเย็น สร้างคอลัมภ์ใหม่

# Assuming 'data' DataFrame is already loaded as shown in your provided code
import pandas as pd

# Convert 'InvoiceDate' to datetime objects if it's not already
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

# Extract the hour from the 'InvoiceDate'
data['Hour'] = data['InvoiceDate'].dt.hour

# Create the 'TimeOfDay' column
def categorize_time_of_day(hour):
    if 6 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 18:
        return 'Afternoon'
    else:
        return 'Evening'

data['TimeOfDay'] = data['Hour'].apply(categorize_time_of_day)

data

<ipython-input-72-b7da359bf7f1>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])
<ipython-input-72-b7da359bf7f1>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Hour'] = data['InvoiceDate'].dt.hour
<ipython-input-72-b7da359bf7f1>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country Legion  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom     EU   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom     EU   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU   
...                    ...        ...         ...             ...    ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France     EU   
541905 2011-12-09 12:50:00       2.10     12680.0          France     EU   
541906 2011-12-09 12:50:00       4.15     12680.0          France     EU   
541907 2011-12-09 12:50:00       4.15     12680.0          France     EU   
541908 2011-12-09 12:50:00       4.95     12680.0          France     EU   

        Hour  TimeOfDay  
0          8    Morning  
1          8    Morning  
2          8    Morning  
3          8    Morning  
4          8    Morning  
...      ...        ...  
541904    12  Afternoon  
541905    12  Afternoon  
541906    12  Afternoon  
541907    12  Afternoon  
541908    12  Afternoon  

[541463 rows x 11 columns]

In [74]:
# prompt: select country = United Kingdom'

country = 'United Kingdom' #@param ["United Kingdom", "France", "Australia", "Netherlands", "Germany", "Norway", "EIRE", "Switzerland", "Spain", "Poland", "Portugal", "Italy", "Belgium", "Lithuania", "Japan", "Iceland", "Channel Islands", "Denmark", "Cyprus", "Sweden", "Finland", "Austria", "Greece", "Hong Kong", "Singapore", "Lebanon", "United Arab Emirates", "Israel", "Saudi Arabia", "Czech Republic", "Canada", "Brazil", "USA", "European Community", "Bahrain", "Malta", "RSA", "Unspecified"]

# Now you can use the selected country in your code
print(f"Selected country: {country}")

# Example usage with the selected country
uk_data = data[data['Country'] == country]
print(uk_data.head())

Selected country: United Kingdom
  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country Legion  Hour  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom     EU     8   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU     8   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom     EU     8   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU     8   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     EU     8   

  TimeOfDay  
0   Morning  
1   Morning  
2   Morning  
3   Morning  
4

In [86]:
morning = uk_data[uk_data['TimeOfDay'] == 'Morning']
afternoon = uk_data[uk_data['TimeOfDay'] == 'Afternoon']
evening = uk_data[uk_data['TimeOfDay'] == 'Evening']

In [88]:
tran_morning = morning.groupby('CustomerID')['StockCode'].apply(list)
tran_afternoon = afternoon.groupby('CustomerID')['StockCode'].apply(list)
tran_evening = evening.groupby('CustomerID')['StockCode'].apply(list)

In [90]:
min_sup = 0.055

frequentItemsets = list(apriori(tran_morning,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'20725', '22384'}), support=0.05510809665112336, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '22384'}), confidence=0.05510809665112336, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'22384'}), confidence=0.560344827586207, lift=7.068735017517979), OrderedStatistic(items_base=frozenset({'22384'}), items_add=frozenset({'20725'}), confidence=0.6951871657754011, lift=7.068735017517979)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.058499364137346334, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.058499364137346334, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.8625, lift=10.5970703125), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.71875, lift=10.597070312500001)])


In [93]:
min_sup = 0.07

frequentItemsets = list(apriori(tran_afternoon,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.0714076675446298, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.0714076675446298, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5339168490153172, lift=5.067760758570386), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6777777777777778, lift=5.067760758570387)])


In [97]:
min_sup = 0.04

frequentItemsets = list(apriori(tran_evening,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'23300', '23301'}), support=0.04411764705882353, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'23300', '23301'}), confidence=0.04411764705882353, lift=1.0), OrderedStatistic(items_base=frozenset({'23300'}), items_add=frozenset({'23301'}), confidence=1.0, lift=17.0), OrderedStatistic(items_base=frozenset({'23301'}), items_add=frozenset({'23300'}), confidence=0.7500000000000001, lift=17.0)])
